pinot-query-runtime/src/main/java/org/apache/pinot/query/runtime/timeseries/TimeSeriesExchangeReceiveOperator.java

```diff
@@ -0,0 +1,182 @@
+      // Step-2: Init timeBuckets and ensure they are the same across all servers.
+      TimeSeriesBlock blockToMerge = (TimeSeriesBlock) result;
+      if (timeBuckets == null) {
+        timeBuckets = blockToMerge.getTimeBuckets();
+      } else {
+        Preconditions.checkState(timeBuckets.equals(blockToMerge.getTimeBuckets()),
+            "Found unequal time buckets from server response");
+      }
+      // Step-3: Merge new block with existing block.
+      for (var entry : blockToMerge.getSeriesMap().entrySet()) {
+        long seriesHash = entry.getKey();
+        List<TimeSeries> currentSeriesList = entry.getValue();
+        TimeSeries sampledTimeSeries = currentSeriesList.get(0);
+        // Init seriesBuilder if required
+        BaseTimeSeriesBuilder seriesBuilder = seriesBuilderMap.get(seriesHash);
+        if (seriesBuilder == null) {
+          seriesBuilder = _factory.newTimeSeriesBuilder(
+              _aggInfo, Long.toString(seriesHash), timeBuckets, sampledTimeSeries.getTagNames(),
+              sampledTimeSeries.getTagValues());
+          seriesBuilderMap.put(seriesHash, seriesBuilder);
+        }
+        for (TimeSeries timeSeries : currentSeriesList) {
+          seriesBuilder.mergeAlignedSeries(timeSeries);
+        }
+      }
+    }
+    // Convert series builders to series and return.
+    Map<Long, List<TimeSeries>> seriesMap = new HashMap<>(seriesBuilderMap.size());
+    for (var entry : seriesBuilderMap.entrySet()) {
+      long seriesHash = entry.getKey();
+      List<TimeSeries> timeSeriesList = new ArrayList<>();
+      timeSeriesList.add(entry.getValue().build());
+      seriesMap.put(seriesHash, timeSeriesList);
+    }
+    return new TimeSeriesBlock(timeBuckets, seriesMap);
+  }
+
+  private TimeSeriesBlock getNextBlockNoAggregation()
+      throws Throwable {
+    Map<Long, List<TimeSeries>> timeSeriesMap = new HashMap<>();
+    TimeBuckets timeBuckets = null;
+    for (int index = 0; index < _numServersQueried; index++) {
+      long remainingTimeMs = _deadlineMs - System.currentTimeMillis();
+      Preconditions.checkState(remainingTimeMs > 0, "Timed out before polling exchange receive");
+      Object result = _receiver.poll(remainingTimeMs, TimeUnit.MILLISECONDS);
+      Preconditions.checkNotNull(result, "Timed out waiting for response. Waited: %s ms", remainingTimeMs);
+      if (result instanceof Throwable) {
+        throw ((Throwable) result);
+      }
+      Preconditions.checkState(result instanceof TimeSeriesBlock,
+          "Found unexpected object. This is a bug: %s", result.getClass());
+      TimeSeriesBlock blockToMerge = (TimeSeriesBlock) result;
+      if (timeBuckets == null) {
+        timeBuckets = blockToMerge.getTimeBuckets();
+      } else {
+        Preconditions.checkState(timeBuckets.equals(blockToMerge.getTimeBuckets()),
+            "Found unequal time buckets from server response");
+      }
+      for (var entry : blockToMerge.getSeriesMap().entrySet()) {
```

Comment:
Will the union of series looks like below?

SeriesHash -----List of [seriesBlockServer1, seriesBlockServer2, ....]

Reply:
let's say we have two servers, they are sending:

```
server-1

seriesHash0 ==> [TimeSeries-1, TimeSeries-2]
seriesHash1 ==> [TimeSeries-3]
```

```
server-2:

seriesHash0 ==> [TimeSeries-4]
```

Then if `getNextBlockNoAggregation` is called, then we'll have:

```
seriesHash0 ==> [TimeSeries-1, TimeSeries-2, TimeSeries-4]
seriesHash1 ==> [TimeSeries-3]
```

You can also refer to the UTs which have some examples.